### PR TITLE
Pre- and Post-Condition Work

### PR DESCRIPTION
**Pull request by romu42**
- [x] signer add process needs to be looked at.
- [x] action and post condition for leave_parent_ds_synced needs to be looked at and discussed.

Note: This is not ready for merge.

### CODE REVIEW

File: fsm/join_sync_dnskeys.go [hunk 1]
```diff
@@ -28,97 +28,6 @@ var FsmJoinSyncDnskeys = music.FSMTransition{
 //      for the next action? I think so. I.e. this implementation (VerifyDnskeysSynched) is
 //      extremely similar to the JoinAddCdsPreCondition function that is the PreCondition for
 //      the next step (adding CDS/CDNSKEYs).
-func VerifyDnskeysSynched(z *music.Zone) bool {
```
Comment (romu42) [inline fsm/join_sync_dnskeys.go:31]: The entire function has just been moved down under the action, no changes were intentionally made to the code

**Comment by johanix** [comment]
Jag ser en massa kod, men jag erkänner att jag inte riktigt tycker mig ha bakgrunden klar för mig. Kan du beskriva vilket problem det är som du håller på att lösa?

Bara "signer removal" låter otäckt :-)

File: fsm/leave_add_cds.go [hunk 2]
```diff
@@ -17,7 +17,7 @@ var FsmLeaveAddCDS = music.FSMTransition{
 
 	PreCondition:  LeaveAddCDSPreCondition,
 	Action:        LeaveAddCDSAction,
-	PostCondition: func(z *music.Zone) bool { return true },
+	PostCondition: LeaveCDSVerify,
```
Comment (johanix) [inline fsm/leave_add_cds.go:20]: If the pre-condition is called LeaveAddCDS... and the action is called LeaveAddCDS..., why isn't the post-condition called LeaveAddCDS...?

File: fsm/leave_parent_ds_synced.go [hunk 3]
```diff
@@ -17,7 +17,7 @@ var FsmLeaveParentDsSynced = music.FSMTransition{
 
 	PreCondition:  LeaveParentDsSyncedPreCondition,
 	Action:        LeaveParentDsSyncedAction,
-	PostCondition: func(z *music.Zone) bool { return true }, // XXX: TODO
+	PostCondition: LeaveVerifyCDSRemoval,
```
Comment (johanix) [inline fsm/leave_parent_ds_synced.go:20]: Ok, so this one is called LeaveVerifyCDS*Removal*, but the other one is called LeaveCDSVerify (no "Add" anywhere). I'd like to have the tokens "Verify" and "CDS" in the same order everywhere and I'd like to have symmetry in naming between Add and Remove functions.

File: fsm/leave_parent_ns_synced.go [hunk 4]
```diff
@@ -197,3 +197,27 @@ func LeaveParentNsSyncedAction(z *music.Zone) bool {
 
 	return true
 }
+
+func LeaveParentNsSyncedConfirmCsyncRemoval(zone *music.Zone) bool {
```
Comment (johanix) [inline fsm/leave_parent_ns_synced.go:201]: Just no.

You've heard the story of the guy that invented a filesystem with infinite compression? Any file could be compressed to a single byte. Only problem was that it required the filesystem to support arbitrarily long filenames...

File: fsm/leave_sync_nses.go [hunk 5]
```diff
@@ -68,7 +68,7 @@ func LeaveSyncNsesAction(z *music.Zone) bool {
 		return false
 	}
 
-	nsrem := []dns.RR{}
+	var nsrem []dns.RR
```
Comment (johanix) [inline fsm/leave_sync_nses.go:71]: Ok, so you don't like to instantiate a new variable with an assignment. Does this mean (I'm really curious) that you're against constructs like:

foo := map[string]*dns.RR{}

and would rather have:

var foo map[string]*dns.RR

?

File: fsm/leave_parent_ds_synced.go (hunk 3 above)
Comment (romu42) [reply fsm/leave_parent_ds_synced.go:20]: Good point.

File: fsm/leave_parent_ns_synced.go (hunk 4 above)
Comment (romu42) [reply fsm/leave_parent_ns_synced.go:201]: True story. I am not happy with that one myself.  The problem is that we have some naming issues throughout MUSIC. I would be more that happy to address these and take the time to decide on how we want it to be.

File: fsm/leave_sync_nses.go (hunk 5 above)
Comment (romu42) [reply fsm/leave_sync_nses.go:71]: My understanding that is better to declare an empty slice with:
` var myslice  []stuff` 
than with : 
`myslice :=  []stuff{}`
according to this: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

I don't think this applies to Maps.

So this isn't a personal opinion as much as me trying to write better go code.

File: fsm/leave_parent_ns_synced.go (hunk 4 above)
Comment (johanix) [reply fsm/leave_parent_ns_synced.go:201]: You know my view: trying to document via long variable names is not the way to go. In this case we're talking about a function that is called from perhaps two places in total. Then a better solution is comments in the code.

If an identifier name is used in a gazillion places (like the discussion about "z" vs "zone") then it is more important to find the "right" name of the identifier.

File: fsm/leave_sync_nses.go (hunk 5 above)
Comment (johanix) [reply fsm/leave_sync_nses.go:71]: Why is it better? Both are legal constructs. And using assignment means we use the same construct for both slices and maps, which makes the code more uniform. That said, my objection is not about the construct

var foo []stuff

but spending time on changing one perfectly legal contruct for another perfectly legal construct.

**Review by johanix — APPROVED** [review]

**Pull request merged**